OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs

```diff
@@ -0,0 +1,327 @@
+		public readonly int[] PaticleSize = { 1, 3 };
+
+		[Desc("Scatters falling direction on the x-axis. Scatter min. and max. value.")]
+		public readonly int[] ScatterDirection = { -1, 1 };
+
+		[Desc("The value that particles fall down. Gravity min. and max. value.")]
+		public readonly int[] Gravity = { 100, 200 };
+
+		[Desc("The current offset value for the swing movement. SwingOffset min. and max. value.")]
+		public readonly int[] SwingOffset = { 10, 15 };
+
+		[Desc("The value that particles swing to the side each update. SwingSpeed min. and max. value.")]
+		public readonly int[] SwingSpeed = { 1, 25 };
+
+		[Desc("The value range that can be swung to the left or right. SwingAmplitude min. and max. value.")]
+		public readonly int[] SwingAmplitude = { 10, 15 };
+
+		[Desc("The colors for the particles. Think about using the colors of the temperate.pal for example.")]
+		public readonly Color[] ParticleColors = {
+			Color.FromArgb(255, 236, 236, 236),
+			Color.FromArgb(255, 228, 228, 228),
+			Color.FromArgb(255, 208, 208, 208),
+			Color.FromArgb(255, 188, 188, 188)
+		};
+
+		[Desc("Works only with line enabled and will fade out the tail of the line.")]
+		public readonly byte LineTailAlphaValue = 200;
+
+		public object Create(ActorInitializer init) { return new WeatherOverlay(init.World, this); }
+	}
+
+	public class WeatherOverlay : ITick, IPostRender
+	{
+		readonly WeatherOverlayInfo info;
+		readonly World world;
+		public struct Particle
+		{
+			public float PosX;
+			public float PosY;
+			public int Size;
+			public float DirectionScatterX;
+			public float Gravity;
+			public float SwingOffset;
+			public float SwingSpeed;
+			public int SwingDirection;
+			public float SwingAmplitude;
+			public Color Color;
+			public Color TailColor;
+		}
+
+		public List<Particle> ParticleList = new List<Particle>();
+		int maxParticleCount;
+
+		enum ParticleCountFaderType { Hold, FadeIn, FadeOut }
+		ParticleCountFaderType particleCountFader = ParticleCountFaderType.FadeIn;
+
+		float targetWindXOffset = 0f;
+		float currentWindXOffset = 0f;
+		int currentWindIndex = 0;
+		long windTickCountdown = 1500;
+		float2 antiScrollPrevTopLeft;
+
+		public WeatherOverlay(World world, WeatherOverlayInfo info)
+		{
+			this.info = info;
+			this.world = world;
+			currentWindIndex = info.WindStrength.Length / 2;
+			targetWindXOffset = info.WindStrength[0];
+			maxParticleCount = CalculateParticleCount(Game.Renderer.Resolution.Width, Game.Renderer.Resolution.Height);
+		}
+
+		int CalculateParticleCount(int x, int y)
+		{
+			return (x * y) / (info.ParticleDensityFactor != 0 ? info.ParticleDensityFactor : 1);
+		}
+
+		void SpawnParticles(int count, int rangeY, int spawnChancePercent)
+		{
+			for (var i = 0; i < count; i++)
+			{
+				if (Game.CosmeticRandom.Next(100) <= spawnChancePercent)
+				{
+					var tempColor = info.ParticleColors[Game.CosmeticRandom.Next(info.ParticleColors.Length)];
```

Comment:
We have specific functions for picking an element in a collection,
IIRC it is something like info.ParticleColors.Random(Game.CosmeticRandom)